contexts/ProposalMarketsContext.tsx

```diff
@@ -0,0 +1,602 @@
+            const _bids = orderBookSide(bids, true);
+            const _asks = orderBookSide(asks);
+            const tobAsk: number = Number(_asks[0][0]);
+            const tobBid: number = Number(_bids[0][0]);
+            return {
+                topAsk: tobAsk,
+                topBid: tobBid,
+            };
+        };
+
+        const getSpreadString = (bids: LeafNode[], asks: LeafNode[]) => {
+            const { topAsk, topBid } = getToB(bids, asks);
+            const spread: number = topAsk - topBid;
+            const spreadPercent: string = ((spread / topBid) * 100).toFixed(2);
+
+            return spread === topAsk
+                ? '∞ (100.00%)'
+                : `${spread.toFixed(2).toString()} (${spreadPercent}%)`;
+        };
+
+        if (markets) {
+            return {
+                passBidsProcessed: getSide(markets.passBids, true),
+                passAsksProcessed: getSide(markets.passAsks),
+                passBidsArray: orderBookSide(markets.passBids, true),
+                passAsksArray: orderBookSide(markets.passAsks),
+                failBidsProcessed: getSide(markets.failBids, true),
+                failAsksProcessed: getSide(markets.failAsks),
+                failBidsArray: orderBookSide(markets.failBids, true),
+                failAsksArray: orderBookSide(markets.failAsks),
+                passToB: getToB(markets.passBids, markets.passAsks),
+                failToB: getToB(markets.failBids, markets.failAsks),
+                passSpreadString: getSpreadString(markets.passBids, markets.passAsks),
+                failSpreadString: getSpreadString(markets.failBids, markets.failAsks),
+            };
+        }
+        return undefined;
+    }, [markets]);
+
+    const placeOrder = useCallback(
+        async (amount: number, price: number, limitOrder?: boolean, ask?: boolean, pass?: boolean) => {
+            if (!proposal || !markets) return;
+            const market = pass
+                ? { publicKey: proposal?.account.openbookPassMarket, account: markets?.pass }
+                : { publicKey: proposal?.account.openbookFailMarket, account: markets?.fail };
+            const placeTxs = await placeOrderTransactions(amount, price, market, limitOrder, ask, pass);
+
+            if (!placeTxs || !wallet.publicKey) {
+                return;
+            }
+
+            try {
+                setLoading(true);
+
+                await sender.send(placeTxs);
+                await fetchMarketsInfo();
+                await fetchOpenOrders(wallet.publicKey);
+            } catch (err) {
+                console.error(err);
+            } finally {
+                setLoading(false);
+            }
+        },
+        [
+            wallet,
+            proposal,
+            markets,
+            connection,
+            sender,
+            placeOrderTransactions,
+            fetchMarketsInfo,
+            fetchOpenOrders,
+        ],
+    );
+
+
+    const consumeOrderBookSide = (
```

Comment:
I think this makes more sense from a function standpoint for getting our data from order book side but likely want some functions wrapping these parsings so we could call them on any bit of data regardless of how we get it (see note above about breaking out toB and spread and whatever). But will leave however much we want to do, open. In that these are more important to NOTE rather than do. As this is just one way to manage books.

Reply:
Ah yeah we definitely could use some helper functions for parsing the order book data... I think I could definitely add some like that down the road? I see it would be helpful with using between both the proposal markets and the open book markets, since that logic is duped now. Do we want to create like a separate issue to add some util/helper functions, or do we want to add that to this PR?